File: utils/src/main/java/uk/chaoticgoose/adventofcode/utils/grid2d/Position2D.java
package uk.chaoticgoose.adventofcode.utils.grid2d;

public record Position2D(Point2D point, Direction2D direction) {

    public Position2D ahead() {
        return new Position2D(point.toThe(direction), direction);
    }

    public Position2D before() {
        return new Position2D(point.toThe(direction.opposite()), direction);
    }

    public Position2D turned90() {
        return new Position2D(point, direction.rotated90());
    }

    public Position2D turned270() {
        return new Position2D(point, direction.rotated270());
    }

    public Position2D turnTo(Direction2D direction) {
        return new Position2D(point, direction);
    }
}

File: twentythree/src/main/kotlin/utils/grids/twodee/Position2D.kt
package utils.grids.twodee

data class Position2D(val point: Point2D, val direction: Direction2D) {
    fun ahead(): Position2D = Position2D(point.toThe(direction), direction)

    fun turnedRight90(): Position2D = Position2D(point, direction.right90())

    fun turnedLeft90(): Position2D = Position2D(point, direction.left90())
}
